### PR TITLE
Add FilePicker component

### PR DESCRIPTION
**Pull request by bytasv**
- Closes https://github.com/mui/mui-toolpad/issues/1395

https://user-images.githubusercontent.com/437214/211545392-37bc80c0-074c-419c-9098-3bd9f805a73f.mp4

### CODE REVIEW

File: packages/toolpad-components/src/FilePicker.tsx
Comment (bytasv) [inline packages/toolpad-components/src/FilePicker.tsx:34]: @Janpot before I sink more time into this I have a few questions:

1. Do you have any alternative suggestions how we should be passing file value to the user? Right now I'm converting `FileList` object to array which should contain actual file value so we can avoid forcing users to use `FileReader` when binding values. At the moment `FileReader` is not even accessible, but also it would require handling promises in binding editor, which I'm not sure if is supported now?
2. This change handler feels a bit ugly as it's handling `onload` and asynchronously setting values on objects. Seems that up until now we didn't need to handle async value setting via change handlers, wondering if we should consider adding `Promise` handling or keep it for sync values only?
3. Without an actual usecase for file picker it's hard to tell what we need, but wondering if dataURL and array buffer would be enough for now? Or maybe I should use something entirely different here?

Comment (Janpot) [reply packages/toolpad-components/src/FilePicker.tsx]: Similar to what I suggested with the DatePicker, remove `onChangeHandler`, program it in the component itself.

Something along the lines of
```tsx
function FilePicker({ multiple, onFilesChange, ...props }: Props) {
  const handleChange = React.useCallback(async (event) => {
    const files = await Promise.all(Array.from(event.target.files, async (file) => {
      return {
        name: file.name,
        type: file.type,
        base64: await readAsBase64(file)
      }
    }))
    onFilesChange(files)
  }, [onFilesChange]);
  
  return <MuiTextField {...props} type="file" onChange={handleChange} inputProps={{ multiple }} />;
}
```

Comment (bytasv) [reply packages/toolpad-components/src/FilePicker.tsx]: Thanks a lot! I didn't got it the first time (with DatePicker), but now it makes sense 👍

File: packages/toolpad-components/src/FilePicker.tsx
Comment (Janpot) [inline packages/toolpad-components/src/FilePicker.tsx:66]:
```suggestion
      typeDef: { type: 'object' },
      visible: false
```

File: packages/toolpad-core/src/types.ts
Comment (Janpot) [inline packages/toolpad-core/src/types.ts:78]: No need for a new type, this is an object. What we need is more power in the object type to represent a shape

**Comment by Janpot** [comment]
Can you also add `helperText` properties to the component and to each arg type definition. See the other components for an example.

**Comment by bytasv** [comment]
@Janpot helper text added

> Also, should we put a bit of effort in making this look a bit better? It feels a bit out of placed compared to the MUI components. Perhaps start by making a benchmark of Material Design (visually) compatible Filepicker components that have been created in the wild?

Part of this PR - I don't think so. I don't see this being a blocker for the functionality. It does however inherit props from `TextField` so it isn't entirely "out of place".
However I do agree that it should look better, though question is whether we should do a special implementation for toolpad only or cooperate with core and finally have file picker component in core library?

**Comment by apedroferreira** [comment]
> -edit: alternatively I could try to use a `Button` component so it looks like a button to select files and not like an input

Sounds like a good idea to use an MUI button at least!

**Comment by bytasv** [comment]
@apedroferreira replaced implementation with button

CleanShot 2023-01-17 at 12 59 29

**Comment by Janpot** [comment]
> Also with an Upload icon if there is one.

Some more options:
* https://mui.com/material-ui/material-icons/?query=file&selected=AttachFile
* https://mui.com/material-ui/material-icons/?query=file&selected=FileOpen
* Maybe there are better to find in https://mui.com/material-ui/material-icons/?query=file ?

And perhaps also show a list of file names that have been selected?

> However I do agree that it should look better, though question is whether we should do a special implementation for toolpad only or cooperate with core and finally have file picker component in core library?

We can always start an implementation and promote it to core later. In any case, I can't seem to find a MUI core or X feature request for a file input, and neither anything in notion (@joserodolfofreitas Would something like this make sense for X at some point?). Maybe step one is to start tracking the demand and benchmark potential solutions?

e.g. a quick google search
- https://codepen.io/Josantonius/pen/BzNwXz
- https://dev.to/j_sakamoto/how-to-implement-a-material-designed-input-type-file-field-built-on-mudblazor-640
- https://codepen.io/akhmedgaziev/pen/JdVQQm
- https://merlosy.github.io/ngx-material-file-input/
- interestingly, it doesn't seem to bring up any with a file drop zone

**Comment by bytasv** [comment]
> And perhaps also show a list of file names that have been selected?

Right in canvas? I'd leave that to the end user to decide, by using other components

> Also with an Upload icon if there is one.

Is the suggestion to hardcode icon along side the `content`?

**Comment by Janpot** [comment]
> Right in canvas? I'd leave that to the end user to decide, by using other components

Yes, that would be just like the native file input, so I don't think it's an unreasonable functionality. We're a low-code tool for internal applications, so our default functionality should cover the majority of use-cases.

> Is the suggestion to hardcode icon along side the content?

The user should get a good default experience. We can allow overriding behavior, but the default should look good and cover the majority of use-cases.

**Comment by bytasv** [comment]
> > Right in canvas? I'd leave that to the end user to decide, by using other components
> 
> Yes, that would be just like the native file input, so I don't think it's an unreasonable functionality. We're a low-code tool for internal applications, so our default functionality should cover the majority of use-cases.
> 
> > Is the suggestion to hardcode icon along side the content?
> 
> The user should get a good default experience. We can allow overriding behavior, but the default should look good and cover the majority of use-cases.

Can you define exactly what you want to see?

I had a default file input behaviour implemented which was just changed to the button, it did show file name out of the box, it also shown if multiple files were selected.

At this point it's not really clear how it should look like.

**Comment by Janpot** [comment]
> At this point it's not really clear how it should look like.

Is it possible to make it look just like the native file input, but with an MUI Button (that has an icon for extra visual feedback)?

**Comment by bytasv** [comment]
Of course it's possible, my question is just, what's the value in current stage of "over optimising" this component so much? If we want default behaviour then why not stick with previous implementation of textfield?

**Comment by joserodolfofreitas** [comment]
> Would something like this make sense for X at some point?

Yes. If you're willing to get started with it, it'd be great. We can incorporate it into MUI X later.

We have an [issue](https://github.com/mui/mui-x/issues/2901) on our repository. And we had a couple of requests for a FilePicker/ Dropzone component in the 2022 survey.
However, at the moment, and with the current bandwidth, it's unlikely that we can undertake this task in 2023. 
The initial plan is to expand the components portfolio with charts and maybe advanced TreeView this year.

**Comment by Janpot** [comment]
> what's the value in current stage of "over optimising" this component so much?

I don't think we're over optimizing here, any PR comment can be dismissed by using the term "over optimizing". The expectation was to visually tie in this component with the rest of the components, ideally without sacrificing too much functionality.

@joserodolfofreitas Thanks, I completely missed that ticket. Been searching on "File input" and "File picker" and "Filepicker" but it doesn't seem to appear on the first few pages. (Makes me think we could introduce a keywords section in the issue template, but that's besides the question). We'll create a basic implementation of a FileInput for toolpad, and we can promote to X when there's more demand.

**Comment by bytasv** [comment]
> I don't think we're over optimizing here

And I think differently. I already had a version done which shows by default file name that's selected and uses default MUI TextField styles.

I can revert changes to that or keep this implementation that uses Button style. What matters is that actual file pick/upload works, and without an actual design or benchmark I don't see the point in optimising this without any external feedback. This can be further improved/optimised if you (or we as entire team) see the benefit.

> any PR comment can be dismissed by using the term "over optimizing"

Any PR that CAN be dismissed by any reasonable argument should be dismissed, this way we can save time and focus on things that matter and that have positive impact.

So in order to stop bikeshedding - which of the 2 options you want me to keep?

**Comment by Janpot** [comment]
There are two different concerns in tension here:

1. We'd like to adhere mostly to a single visual language across our components so that our users can create coherent UIs
2. A component should visually represent the state it is in. e.g. a text input or select shows the actual value, even when it's not being interacted with.

- The default file input gives us 2.
- The Button implementation gives us 1.
- My proposal would be to accompany that Button with a Typography to give use 1. and 2.

I understand that this would be some extra work, so if I really had to choose, I'd say to use the default file input. It has the better UX for the user, and is easier to improve for us afterwards.

Before we merge, if you're up for it, I'd propose to add a simple test, like we have for all the other components.

**Comment by bytasv** [comment]
This is how the 1st option looked like pretty much:

https://codesandbox.io/s/red-water-uqq1po?file=/demo.tsx

CleanShot 2023-01-17 at 19 49 08@2x

So saying that "it does not adhere to single visual language" is a bit of exaggeration, don't you think? While I agree it can be further OPTIMISED by getting rid of default browser button, I don't see value at this point as per comments explained before - there is no design (which means its likely going to change anyway) and we haven't done any benchmark (which means its likely to change anyway).

> Before we merge, if you're up for it, I'd propose to add a basic test, like we have for all the other components.

Sure I will try to add tests only if playwright allows working with file inputs 🤔

**Comment by Janpot** [comment]
> So saying that "it does not adhere to single visual language" is a bit of exaggeration, don't you think?

👍  Ok, I see now where our misunderstanding originates from. For me, having a browser native button nested inside of a material design textfield would be the definition of "it does not adhere to single visual language". But I understand that the threshold is interpreted differently by different people.

**Comment by bytasv** [comment]
reverted to textfield version and added test, please review

**Comment by Janpot** [comment]
There's one thought I had which may be a bit of a nit, but is related to updating this to a proper FilePicker component in the future. I can imagine that new component won't have MUI `TextField` specific properties like `variant`, `size` or `fullWidth`. It could be tricky to migrate if we let users build apps with these properties and have to upgrade this component to a version that doesn't have them. Would it make sense to omit those properties for now, so that the next programmer working on this code has an easier job?

**Comment by bytasv** [comment]
Props removed

**Review by Janpot — APPROVED** [review]

**Pull request merged**
